### ISSUE TITLE
TKECERT: Extension load failure with release 1.8

### ISSUE THREAD

**Issue by githje** [sassoftware/python-swat#106]
Hi,

using the latest 1.8 release gives me this error on a Python 3.8.6 environment:

```
RuntimeError: Booting the TK subsystem failed: TKECERT: Extension load failure
```

Happens for both methods (binary and REST). Working fine with the 1.7.1 release on the same environment. I'm using the Jupyter docker image downloaded from here:
[https://hub.docker.com/r/jupyter/datascience-notebook/](https://hub.docker.com/r/jupyter/datascience-notebook/)

Here's the full output:
```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-1-89354108e34b> in <module>
----> 1 import swat
      2 
      3 # http/REST connection method
      4 sess = swat.CAS("mymachine", 8777, "myuser", "mypasswd")
      5 

/opt/conda/lib/python3.8/site-packages/swat/__init__.py in <module>
     70 
     71 # Configuration
---> 72 from . import config    # noqa: E402
     73 from .config import (set_option, get_option, reset_option, describe_option,
     74                      options, option_context)    # noqa: E402

/opt/conda/lib/python3.8/site-packages/swat/config.py in <module>
     81 
     82 
---> 83 register_option('tkpath', 'string', set_tkpath, _initialize_tkpath(),
     84                 'Displays the path for SAS TK components.  This is determined\n'
     85                 'when SWAT is imported.  By default, it points to the platform\n'

/opt/conda/lib/python3.8/site-packages/swat/utils/config.py in register_option(key, typedesc, validator, default, doc, environ)
    762     with warnings.catch_warnings():
    763         warnings.simplefilter('ignore')
--> 764         _config[key] = SWATOption(key, typedesc, validator, default, doc, environ=environ)
    765 
    766 

/opt/conda/lib/python3.8/site-packages/swat/utils/config.py in __init__(self, name, typedesc, validator, default, doc, environ)
    654             self._default = validator(_getenv(environ, default))
    655         else:
--> 656             self._default = validator(default)
    657         self._environ = environ
    658         self._value = self._default

/opt/conda/lib/python3.8/site-packages/swat/config.py in set_tkpath(val)
     52         return
     53     path = check_string(val)
---> 54     InitializeTK(a2n(path, 'utf-8'))
     55     return path
     56 

/opt/conda/lib/python3.8/site-packages/swat/clib.py in InitializeTK(*args, **kwargs)
    163 
    164     try:
--> 165         out = _pyswat.InitializeTK(*args, **kwargs)
    166 
    167     finally:

RuntimeError: Booting the TK subsystem failed: TKECERT: Extension load failure.
```

**Comment by kesmit13** [comment]
I'm not able to reproduce this issue using Docker. I've created a new container using the following Dockerfile and it works. Are there additional steps that you are doing that might be affecting the behavior?
```
FROM jupyter/datascience-notebook:latest
RUN conda install -c sas-institute swat==1.8.0
```

**Comment by githje** [comment]
No, actually my Dockerfile is as simple as yours:
```
FROM jupyter/datascience-notebook:python-3.8.6
RUN pip install 'swat'
```
I'm not using "latest" and I'm using pip. Feel free to close if you can't reproduce. I'm not depending on the 1.8 release and will try again with the next release.

**Comment by kesmit13** [comment]
Hmm, I tried that Dockerfile too and it's working from Docker on OS X.

**Comment by githje** [comment]
I tried again and it still happens. This time I submitted the code from the command-line Python inside the container and got a more detailed error message:

```
>>> import swat
ERROR: Could not load opensslcert extension. OpenSSL libraries are not found or the wrong version.
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/conda/lib/python3.8/site-packages/swat/__init__.py", line 72, in <module>
    from . import config    # noqa: E402
  File "/opt/conda/lib/python3.8/site-packages/swat/config.py", line 83, in <module>
    register_option('tkpath', 'string', set_tkpath, _initialize_tkpath(),
  File "/opt/conda/lib/python3.8/site-packages/swat/utils/config.py", line 764, in register_option
    _config[key] = SWATOption(key, typedesc, validator, default, doc, environ=environ)
  File "/opt/conda/lib/python3.8/site-packages/swat/utils/config.py", line 656, in __init__
    self._default = validator(default)
  File "/opt/conda/lib/python3.8/site-packages/swat/config.py", line 54, in set_tkpath
    InitializeTK(a2n(path, 'utf-8'))
  File "/opt/conda/lib/python3.8/site-packages/swat/clib.py", line 165, in InitializeTK
    out = _pyswat.InitializeTK(*args, **kwargs)
RuntimeError: Booting the TK subsystem failed: TKECERT: Extension load failure.
```

I compared the opensslcert*.so files between both containers (SWAT 1.7.1 and SWAT 1.8). They are not identical:

SWAT 1.7.1 (ok):
```
$ ll /opt/conda/lib/python3.8/site-packages/swat/lib/linux/opensslcert*
-rw-r--r--. 1 jovyan users 156861 Jan 14 12:19 /opt/conda/lib/python3.8/site-packages/swat/lib/linux/opensslcert100.so
-rw-r--r--. 1 jovyan users 156861 Jan 14 12:19 /opt/conda/lib/python3.8/site-packages/swat/lib/linux/opensslcert101.so
-rw-r--r--. 1 jovyan users 156861 Jan 14 12:19 /opt/conda/lib/python3.8/site-packages/swat/lib/linux/opensslcert102.so
-rw-r--r--. 1 jovyan users 156893 Jan 14 12:19 /opt/conda/lib/python3.8/site-packages/swat/lib/linux/opensslcert110.so
-rw-r--r--. 1 jovyan users 148777 Jan 14 12:19 /opt/conda/lib/python3.8/site-packages/swat/lib/linux/opensslcert111.so
```

SWAT 1.8 (not ok):
```
$ ll /opt/conda/lib/python3.8/site-packages/swat/lib/linux/opensslcert*
-rw-r--r--. 1 jovyan users 198739 Jan 19 17:38 /opt/conda/lib/python3.8/site-packages/swat/lib/linux/opensslcert100.so
-rw-r--r--. 1 jovyan users 198739 Jan 19 17:38 /opt/conda/lib/python3.8/site-packages/swat/lib/linux/opensslcert101.so
-rw-r--r--. 1 jovyan users 198739 Jan 19 17:38 /opt/conda/lib/python3.8/site-packages/swat/lib/linux/opensslcert102.so
-rw-r--r--. 1 jovyan users 198772 Jan 19 17:38 /opt/conda/lib/python3.8/site-packages/swat/lib/linux/opensslcert110.so
-rw-r--r--. 1 jovyan users 190655 Jan 19 17:38 /opt/conda/lib/python3.8/site-packages/swat/lib/linux/opensslcert111.so
```

Same number of files but different file sizes? Could that explain something?

**Comment by kesmit13** [comment]
Can you do `ldconfig -v | grep ssl` to show what SSL libraries are being used?

**Comment by kesmit13** [comment]
It is possible to set an explicit path to the OpenSSL library using the `TKESSL_OPENSSL_LIB` environment variable. You might try setting that to `/opt/conda/lib/libssl.so.1.1` or `/lib/x86_64-linux-gnu/libssl.so.1.1` and see if that changes anything.

**Comment by githje** [comment]
Unfortunately, no luck. I rebuilt both containers with nearly identical Dockerfiles:

```
FROM jupyter/datascience-notebook:python-3.8.6
RUN pip install 'swat==1.7.1'
USER root
```
and

```
FROM jupyter/datascience-notebook:python-3.8.6
RUN pip install 'swat'
USER root
```

Still have the issue with the 1.8 version. Setting the TKESSL_OPENSSL_LIB var does not help. `/opt/conda/lib/libssl.so.1.1` is identical in both containers.

```
# ll /opt/conda/lib/libssl.so.1.1
-rwxrwxr-x. 1 jovyan users 674312 Dec 11 22:58 /opt/conda/lib/libssl.so.1.1*
````

The other command produced identical output as well:

```
# ldconfig -v | grep ssl
        libssl.so.1.1 -> libssl.so.1.1
        libssl3.so -> libssl3.so
```

One thing I didn't mention before - I run both containers on vanilla Kubernetes (1.18.10) on a CentOS 7.9 worker node. I don't assume that this makes a difference?

**Comment by kesmit13** [comment]
We have been looking into this issue and it does appear that running in k8s is the key element to getting this crash to surface. Running a container locally works fine. We still haven't figured out the source of the problem yet though.

**Comment by kesmit13** [comment]
You can set the following environment variable to fix the problem. This directory is not in the default search path for the C extension.
```
TKECERT_CRYPTO_LIB=/usr/lib/x86_64-linux-gnu/libcrypto.so
```
I'll add some code to SWAT to do this lookup automatically for the next release.

**Comment by githje** [comment]
Awesome! I'll try this as soon as I have access to the cluster again. I'll close the ticket for now. Thanks a lot for your support.

**Comment by kesmit13** [comment]
Fixed in v1.8.1
